### PR TITLE
Simplify ScriptModule and script registration

### PR DESCRIPTION
**Pull request by s1monw**
Registering a script engine or native scripts still uses Guice today
and is much more complicated than needed. This change moves to a pull
based model where script plugins have to implement a dedicated interface
`ScriptPlugin` and defines simple getter returning instances rather than
classes.

### CODE REVIEW

**Comment by s1monw** [comment]
@rjernst can you take a look at this too? @jasontedor you might be interested too though

File: core/src/main/java/org/elasticsearch/plugins/ScriptPlugin.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+     * Returns a {@link ScriptEngineService} instance or <code>null</code> if this plugin doesn't add a new script engine
+     */
+    default ScriptEngineService getScriptEngineService(Settings settings) {
+        return null;
+    }
+
+    /**
+     * Returns a list of {@link NativeScriptFactory} instances.
+     */
+    default List<NativeScriptFactory> getScriptFactories() {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/ScriptPlugin.java]: Maybe call this `getNativeScripts()`? I think that makes it more clear what is returned, vs just following exactly the fact that native scripts are implemented through a factory.

File: core/src/main/java/org/elasticsearch/plugins/ScriptPlugin.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        return Collections.emptyList();
+    }
+
+    /**
+     * Returns a {@link ScriptContext.Plugin} instance or <code>null</code> if this plugin doesn't add a new script context plugin
+     */
+    default ScriptContext.Plugin getScriptContextPlugin() {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/ScriptPlugin.java]: Maybe call this `getCustomScriptContexts()`?

**Comment by rjernst** [comment]
LGTM!

**Comment by s1monw** [comment]
thanks @rjernst

**Pull request merged**